libs/common/src/key-management/pin/pin-state.service.implementation.ts

```diff
@@ -0,0 +1,119 @@
+import {
+  PIN_PROTECTED_USER_KEY_ENVELOPE_PERSISTENT,
+  PIN_PROTECTED_USER_KEY_ENVELOPE_EPHEMERAL,
+  USER_KEY_ENCRYPTED_PIN,
+  PIN_KEY_ENCRYPTED_USER_KEY_PERSISTENT,
+} from "./pin.state";
+
+export class PinStateService implements PinStateServiceAbstraction {
+  constructor(private stateProvider: StateProvider) {}
+
+  getUserKeyWrappedPin$(userId: UserId): Observable<EncString | null> {
+    assertNonNullish(userId, "userId");
+
+    return this.stateProvider
+      .getUserState$(USER_KEY_ENCRYPTED_PIN, userId)
+      .pipe(map((value) => (value ? new EncString(value) : null)));
+  }
+
+  async getPinLockType(userId: UserId): Promise<PinLockType> {
+    assertNonNullish(userId, "userId");
+
+    const isPersistentPinSet =
+      (await this.getPinProtectedUserKeyEnvelope(userId, "PERSISTENT")) != null ||
+      // Deprecated
+      (await this.getLegacyPinKeyEncryptedUserKeyPersistent(userId)) != null;
+    const isPinSet =
+      (await firstValueFrom(this.stateProvider.getUserState$(USER_KEY_ENCRYPTED_PIN, userId))) !=
+      null;
+
+    if (isPersistentPinSet) {
+      return "PERSISTENT";
+    } else if (isPinSet) {
+      return "EPHEMERAL";
+    } else {
+      return "DISABLED";
+    }
+  }
+
+  async getPinProtectedUserKeyEnvelope(
+    userId: UserId,
+    pinLockType: PinLockType,
+  ): Promise<PasswordProtectedKeyEnvelope | null> {
+    assertNonNullish(userId, "userId");
+
+    if (pinLockType === "EPHEMERAL") {
+      return await firstValueFrom(
+        this.stateProvider.getUserState$(PIN_PROTECTED_USER_KEY_ENVELOPE_EPHEMERAL, userId),
+      );
+    } else if (pinLockType === "PERSISTENT") {
+      return await firstValueFrom(
+        this.stateProvider.getUserState$(PIN_PROTECTED_USER_KEY_ENVELOPE_PERSISTENT, userId),
+      );
+    } else {
+      throw new Error(`Unsupported PinLockType: ${pinLockType}`);
+    }
+  }
+
+  async getLegacyPinKeyEncryptedUserKeyPersistent(userId: UserId): Promise<EncString | null> {
+    assertNonNullish(userId, "userId");
+
+    return await firstValueFrom(
+      this.stateProvider
+        .getUserState$(PIN_KEY_ENCRYPTED_USER_KEY_PERSISTENT, userId)
+        .pipe(map((value) => (value ? new EncString(value) : null))),
+    );
+  }
+
+  async setPinState(
+    userId: UserId,
+    pinProtectedUserKeyEnvelope: PasswordProtectedKeyEnvelope,
+    userKeyEncryptedPin: EncryptedString,
+    pinLockType: PinLockType,
+  ): Promise<void> {
+    assertNonNullish(userId, "userId");
+    assertNonNullish(pinProtectedUserKeyEnvelope, "pinProtectedUserKeyEnvelope");
+    assertNonNullish(pinLockType, "pinLockType");
+
+    let keyDefinition;
+    if (pinLockType === "EPHEMERAL") {
+      keyDefinition = PIN_PROTECTED_USER_KEY_ENVELOPE_EPHEMERAL;
+    } else if (pinLockType === "PERSISTENT") {
+      keyDefinition = PIN_PROTECTED_USER_KEY_ENVELOPE_PERSISTENT;
+    } else {
+      throw new Error(`Cannot set up PIN with pin lock type ${pinLockType}`);
+    }
```

Comment:
❓ is `keyDefinition` type `any` ?
Maybe it would be better to just call set it to state, without temporary variable:
```suggestion
    if (pinLockType === "EPHEMERAL") {
      await this.stateProvider.setUserState(PIN_PROTECTED_USER_KEY_ENVELOPE_EPHEMERAL, pinProtectedUserKeyEnvelope, userId);
    } else if (pinLockType === "PERSISTENT") {
      await this.stateProvider.setUserState(PIN_PROTECTED_USER_KEY_ENVELOPE_PERSISTENT, pinProtectedUserKeyEnvelope, userId);
    } else {
      throw new Error(`Cannot set up PIN with pin lock type ${pinLockType}`);
    }
```